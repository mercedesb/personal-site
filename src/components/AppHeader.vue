<template>
  <nav class='MainNav'>
    <div :class="'FlexContainer FlexContainer--alignCenter FlexContainer--' + classModifier">
      <smart-link class="MainNav-home" to="/"><span class='MainNav-logo'>Mercedes Bernard</span></smart-link>
      <div>
        <div class='MainNav-hamburger' v-on:click="expanded = !expanded">
          <img src='../assets/menu.svg' alt='hamburger menu'/>
        </div>
        <div>
          <ul :class="'MainNav-navLinks MainNav-navLinks--' + navStateClass" v-if="navLinks.length">
            <li class="MainNav-navLink" v-for="navLink in navLinks">
              <smart-link
                :to="`${navLink.externalLink || navLink.urlSegment}`"
                :isExternal="!!navLink.externalLink"
                >
                {{navLink.title}}
              </smart-link>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </nav>
</template>

<script>
export default {
  props: {
    classModifier: String
  },
  data() {
    return {
      expanded: false
    }
  },
  computed: {
    navLinks() {
      return this.$store.state.navLinks
    },
    navStateClass() {
      return this.expanded ? 'expanded' : 'collapsed'
    }
  },
  created () {
    this.$store.dispatch('getNavLinks')
  }
}
</script>

<style lang="scss" scoped>
  $nav-height: 75px;
  $hamburger-width: 300px;

  .MainNav {
    background-color: $white;
    color: $black;
    box-shadow: $base-drop-shadow;
    z-index: 3;

    &-home {
      display: flex;
      align-items: center;
      padding: $small-spacing $base-spacing;
    }

    &-logo {
      max-height: 70px;
      font-family: $candela;
      font-size: $larger-font-size;
      font-weight: $heavy-font-weight;
    }

    &-hamburger {
      margin-right: $base-spacing;
      justify-self: flex-end;
      img {
        width: $nav-height - $base-spacing;
      }
    }

    &-navLinks {
      position: absolute;
      background-color: $white;
      width: $hamburger-width;
      right: 0;
      display: flex;
      flex-direction: column;
      box-shadow: 0px 10px 10px $black, -10px 8px 10px $black, 0px 10px 10px $black;

      &--collapsed {
        display:none;
      }
    }

    &-navLink {
      padding: $small-spacing $base-spacing;
      text-align: right;
    }

    .FlexContainer--justifyCenter {
      .MainNav-hamburger, .MainNav-navLinks {
        display:none;
      }
    }

    .FlexContainer--spaceBetween {
      .MainNav-link {
        padding-right: $base-spacing;
      }
    }
  }
</style>