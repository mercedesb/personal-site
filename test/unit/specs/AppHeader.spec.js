import AppHeader from '@/components/AppHeader.vue'
import { TestUtility } from './TestUtility'

describe('AppHeader', () => {
  let component

  const initialProps = {
    classModifier: 'modifier'
  }
  const store = {
    state: {
      navLinks: Object.keys(TestUtility.landingPages).map((key) => {
        return TestUtility.landingPages[key]
      })
    },
    actions: {
      getNavLinks: jest.fn()
    }
  }

  const shallow = propsData => TestUtility.shallow(AppHeader, {
    store,
    propsData: {
      ...initialProps,
      ...propsData
    }
  })

  describe('Snapshots', () => {
    describe('with a classModifier passed in', () => {
      it('matches snapshot', () => {
        component = shallow()
        expect(component.element).toMatchSnapshot()
      })
    })

    describe('without a classModifier passed in', () => {
      it('matches snapshot', () => {
        component = shallow({classModifier: undefined})
        expect(component.element).toMatchSnapshot()
      })
    })
  })

  describe('Properties', () => {
    it('has a classModifier property', () => {
      component = shallow()
      expect(component.props().classModifier).toBe('modifier')
    })
  })

  describe('Computed', () => {
    describe('navLinks', () => {
      beforeEach(() => {
        component = shallow()
      })

      it('has navLinks populated', () => {
        expect(component.vm.navLinks).toBeDefined()
      })

      it('creates a parsedLink prop on the navLinks', () => {
        expect(component.vm.navLinks[0].parsedLink).toBeDefined()
      })

      it('uses the external link for parsedLink if its defined', () => {
        const shopPage = TestUtility.landingPages['shop']
        const shopLink = component.vm.navLinks.find((link) => {
          return link.title === shopPage.title
        })
        expect(shopLink.parsedLink).toEqual(shopPage.externalLink)
      })

      it('uses the url segment for parsedLink if eternalLink is not defined', () => {
        const blogPage = TestUtility.landingPages['blog']
        const blogLink = component.vm.navLinks.find((link) => {
          return link.title === blogPage.title
        })
        expect(blogLink.parsedLink).toEqual(`/${blogPage.urlSegment}`)
      })
    })
  })

  describe('Lifecycle', () => {
    describe('created', () => {
      it('dispatches getNavLinks to the store', () => {
        component = TestUtility.mount(AppHeader, {
          store,
          propsData: {
            ...initialProps
          }
        })
        expect(component.vm.$store.dispatch).toHaveBeenCalledWith('getNavLinks')
      })
    })
  })
})
