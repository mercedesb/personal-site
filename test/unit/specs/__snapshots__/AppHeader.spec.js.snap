// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppHeader Snapshots with a classModifier passed in matches snapshot 1`] = `<!---->`;

exports[`AppHeader Snapshots without a classModifier passed in matches snapshot 1`] = `<!---->`;
